components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,596 @@
+// This file is part of ICU4 and_or: todo!(), polarity: todo!(), operand: todo!(), modulo: todo!(), range_list: todo!()  and_or: todo!(), polarity: todo!(), operand: todo!(), modulo: todo!(), range_list: todo!() X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#![allow(missing_docs)]
+
+use crate::rules::reference;
+use core::{convert::TryInto, fmt, mem, str::FromStr};
+use icu_provider::yoke::{self, *};
+use zerovec::{
+    ule::{custom::EncodeAsVarULE, AsULE, PlainOldULE, VarULE, ULE},
+    {VarZeroVec, ZeroVec},
+};
+
+#[derive(Debug, Yokeable, ZeroCopyFrom)]
+pub struct Rule<'data>(pub VarZeroVec<'data, RelationULE>);
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[repr(u8)]
+pub enum AndOr {
+    Or,
+    And,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[repr(u8)]
+pub enum Polarity {
+    Negative,
+    Positive,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(u8)]
+pub enum Operand {
+    N,
+    I,
+    V,
+    W,
+    F,
+    T,
+    C,
+    E,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RangeOrValue {
+    Range(u32, u32), // XXX: Can we get away from smaller?
+    Value(u32),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Relation<'data> {
+    pub(crate) and_or: AndOr,
+    pub(crate) polarity: Polarity,
```

Comment:
thought (also a note to self, really): if these are refactored into an AndOrPolarityOperand struct on the non-ULE side and we just implement ULE on that (instead of doing the `andor_polarity_operand` field) this becomes something much more amenable to the custom derive that I hope to have